contracts/index_executor/src/traits.rs

```diff
@@ -1,15 +1,38 @@
 use super::context::Context;
+use alloc::vec::Vec;
+use phat_offchain_rollup::clients::substrate::SubstrateRollupClient;
 
 pub trait Runner {
     /// Check if a job can be executed.
     /// If the transaction already sent to blockchain, e.g. can be found in memory pool,
     /// it should be `unrunable`.
     /// If the transaction failed to execute, it should be `unrunable`.
     /// Else the job should be `runnable`.
-    fn runnable(&self) -> bool;
-    /// Execute a job, basically send a transaction to blockchain.
-    fn run(&self, context: &Context) -> Result<(), &'static str>;
-    /// Check if a job is already executed successfully.
+    fn runnable(
+        &self,
+        nonce: u64,
+        context: &Context,
+        client: Option<&mut SubstrateRollupClient>,
+    ) -> Result<bool, &'static str>;
+
+    /// Execute a job, basically send a transaction to blockchain, and return tx id.
+    fn run(
+        &self,
+        nonce: u64,
+        recipient: Option<Vec<u8>>,
```

Comment:
Why make `recipient` Option? Because before the task is initialized, we don't need to set `recipient` for each step. The `recipient` always be the worker account address on the dest chain except last step

Reply:
Why don't make `recipient` an argument in the step data (parameter in `new()`)?

In my understanding, each step roughly represents an action to sent out a tx. `nonce` is a variable that applies to all the tx. So it's easy to understand `nonce` is a parameter of `run()`. The sender account can be found from `context`. However, `recipient` is subject to the actual action. For example, some swap may not allow to set a recipient. So do you think it's more natural to put it in the step inner data?

Reply:
Agree, put it inside step now